Rule: aria-required-owned-element (#245)
act-rules/act-rules.github.io: _rules/aria-required-owned-element-bc4a75.md

## Assumptions

Attributes that start with `aria-*` are used to set either [state](https://www.w3.org/WAI/PF/aria-1.1/terms#def_state) or [property](https://www.w3.org/WAI/PF/aria-1.1/terms#def_property) that should be picked up by assistive technologies. If any `aria-*` attributes are not used for that purpose, the results of this rule may be incorrect.

## Accessibility Support

This rule relies on assistive technologies to recognize owned elements, as defined by [WAI-ARIA 1.1](https://www.w3.org/TR/wai-aria). This includes when they are nested descendants that are not immediate children. However, some assistive technologies do not recognize owned elements that are not immediate children, unless workarounds are used.

## Background

- [Required Owned Element](https://www.w3.org/TR/wai-aria-1.1/#mustContain)
- [Owned Element](https://www.w3.org/TR/wai-aria-1.1/#dfn-owned-element)
- [HTML in ARIA](https://www.w3.org/TR/html-aria/)
- [Implicit WAI-ARIA Semantics](https://www.w3.org/TR/wai-aria-1.1/#implicit_semantics)

## Test Cases

> ### Passed

Reviewer: Some suggestions for further pass examples
- One passed example with aria-busy
- One passed example where the top-level element in the example does not have an explicit role (since it is no longer an applicability requirement)
- One passed example where there are elements with roles that meet the owned by requirements, but also other elements that do not

Author: Added examples for points 1 & 2, as for 3 believe passed example 1 and 4 convey this already.

#### Passed example 1

Element `ul` with role `list` owns at least one instance of required owned element, in this case a `span` with `role` of `listitem`.

```html
<ul role="list">
	<span role="listitem"></span>
	<span></span>
</ul>
```

#### Passed example 2

Element `div` with role `tablist` owns an instance of required owned element, a `span` with `role` of `tab`.

```html
<div role="tablist">
	<span role="tab"> </span>
</div>
```

#### Passed example 3

Elements with multiple nested roles, that each own at least an instance of required owned element.

```html
<table role="grid">
	<tr role="row">
		<span role="cell"> </span>
	</tr>
</table>
```

#### Passed example 4

Multiple types of required owned elements are present.

```html
<ul role="menu">
	<li></li>

> 	<!-- implicit role -->

Reviewer: Should be in the line above

Element with role `table` owns multiple instances of required owned elements. Note that the owned elements do not have to be an immediate children of the target element. (See accessibility support notes for more information).

```html
<div role="table">
	<div>
		<div>
			<div role="row">
				<div role="cell">my cell</div>
				<div role="cell">and another cell</div>
			</div>
			<div role="row">
				<div role="cell">my cell</div>
				<div role="cell">and another cell</div>
			</div>
		</div>
	</div>
</div>
```

#### Passed example 6

Element with role `grid` owns multiple instances of required owned elements.

```html
<div role="grid">
	<div role="rowgroup">
		<div role="row">
			<div role="gridcell">1</div>
			<div role="gridcell">Item 1</div>
			<div role="gridcell">2 Nos.</div>
		</div>
	</div>
	<div role="rowgroup">
		<div role="row">
			<div role="gridcell">1</div>
			<div role="gridcell">Item 1</div>
			<div role="gridcell">2 Nos.</div>
		</div>
	</div>
</div>
```

#### Passed example 7

Element with role `list` owns an instance of required owned element which has an implicit role of `listitem`.

```html
<div role="list">
	<li></li>
</div>
```

#### Passed example 8

> Element with role `menubar` can own multiple required owned element with role `menuitem`, `menuitemcheckbox` and `menuitmeradio`, but only one is necessary.

Reviewer: This description is "strange", since the example shows two "owned roles". With this description, I was expecting just one.

```html
<div role="menubar">
	<div role="menuitem"></div>
	<div role="menuitemcheckbox"></div>
</div>
```

#### Passed example 9

Element with role `list` contains a required `listitem`, the relationship between both elements is made possible through the `aria-owns` properties.

**Note:** This test case follows the definition of [owned by](#owned-by). If implemented differently, this definition could cause differences in outcome of this test case.

```html
<div role="list" aria-owns="id1"></div>
<div id="id1" role="listitem"></div>
```

> ### Failed

Reviewer: One suggestion for another fail example
- One fail example where the top-level element in the example does not have an explicit role (since it is no longer an applicability requirement)

#### Failed example 1

Element with role `list` is missing required owned element `list`.

```html
<ul role="list">
	<span> </span>
</ul>
```

#### Failed example 2

Element with role `tablist` is missing required owned element `tab`.

```html
<ol role="tablist">
	<li role="listitem"></li>
</ol>
```

#### Failed example 3

Element with role `grid` owns required owned element `rowgroup`, which is missing its required owned element `row`.

```html
<div role="grid">
	<div>
		<div role="rowgroup">
			<span>Item 1</span>
		</div>
		<div></div>
	</div>
</div>
```

#### Failed example 4

The element with the semantic role of `list` owns the element with the role `tab` through the `aria-owns` property.

**Note:** This test case follows the definition of [owned by](#owned-by). If implemented differently, this definition could cause differences in outcome of this test case.

```html
<div role="list" aria-owns="id2"></div>
<div id="id2" role="tab"></div>
```

> ### Inapplicable

Reviewer: One suggestion for an inapplicable example
- one element with a role that does not have required owned roles (e.g. region, progressbar)

#### Inapplicable example 1

Element with explicit semantic role `list` is not included in the accessibility tree.

```html
<ul role="list" aria-hidden="true">
	<li></li>
</ul>
```

#### Inapplicable example 2

Element has empty role.

```html
<ul role="">
	<li></li>
</ul>
```

#### Inapplicable example 3

> Element has no explicit semantic role.

Reviewer: This is no longer inapplicable

**Note:** An example of an element that has a [required owned element][] is `tablist` which has `tab` as a [required owned element][].

**Note:** An example of an element that has an [implicit semantic role][] that is identical to its [explicit semantic role][] is a `ul` element that has `role="list"`. These elements are not applicable.

**Note:** The applicability of this rule is limited to only the [WAI-ARIA 1.1][] roles, since there are unresolved issues with how [Digital Publishing WAI-ARIA Module][] (DPUB ARIA 1.1) uses role inheritance to define the [required owned element][], which makes it deviate from the model defined in [WAI-ARIA 1.1][]. The [WAI-ARIA Graphics Module][] does not include any [required owned element][].

**Note:** The combobox role is excluded from this rule, because the design pattern for it as described in ARIA 1.1 has proven problematic. The combobox will be significantly different for ARIA 1.2, where it does not have [required owned elements][].

## Expectation

Each test target only [owns][] elements with a [semantic role][] from the [required owned element][] list for the test target's [semantic role]().

> **Note:** If the [required owned element][] list has two roles in one item separated by a "→", elements with the first of these roles, is only a [required owned element][] if all its owned elements have the second of these roles as their [semantic role][]. For example a `list` has `group → listitem`, so an element with the [semantic role][] of `group` only counts as a [required owned element][] for `list`, if all elements [owned by][] the `group` have the [semantic role][] of `listitem`.

Reviewer 2: I understand what this means, but I had to dig into the specs to get it, and I don't think that this sentence currently makes sense…

```suggestion
**Note:** Some [required owned elements][] are only valid if they themselves [own][owns] (or "contain") elements with a given [semantic role][]. This is denoted by an arrow (meaning "containing") in the role description. For example, the role `list` has `group → listitem` as one of its [required owned elements][], meaning that elements with a role of `list` may only [owns][] elements with a role of `group` who themselves only [own][] elements with a role of `listitem`.
```

(I think it's important to use the word "contain" to denote the arrow, given that it is how ARIA calls it)

- [Required Owned Element](https://www.w3.org/TR/wai-aria-1.1/#mustContain)
- [Owned Element](https://www.w3.org/TR/wai-aria-1.1/#dfn-owned-element)

## Test Cases

### Passed

#### Passed Example 1

This element with the `list` role only owns elements with the `listitem` role. The `listitem` role is one of the [required owned elements][] for `list`.

```html
<div role="list">
	<span role="listitem">Item 1</span>
	<span role="listitem">Item 2</span>
</div>
```

#### Passed Example 2

This element with the `tablist` role only owns elements with the `tab` role. The `tab` role is one of the [required owned elements][] for `tablist`.

```html
<div role="tablist">
	<span role="tab">Tab 1</span>
	<span role="tab">Tab 2</span>
</div>
```

#### Passed Example 3

This element with the `grid` role only owns elements with the `row` role, and the element with the `row` role only owns elements with the `cell` role. The `row` role is one of the [required owned elements][] for `grid`, and `cell` is one of the [required owned elements][] for `row`.

```html
<table role="grid">
	<tr role="row">
		<span role="cell">Item 1</span>
	</tr>
</table>

> ```

Reviewer 2: I think that Passed Examples 1, 2, and 3 are very similar (simple good stuff) with 3 being a bit different (nested roles with required ownership).
I suggest we remove either Passed Example 1 or Passed Example 2.

#### Passed Example 4

> This element with the `menu` role only owns elements with the `menuitem`, `menuitemradio` and `menuitemcheckbox` role. These roles are one of the [required owned elements][] for `menu`. The element with the `none` role is not [owned by][] the `menu` because it is not [included in the accessibility tree][].

Reviewer 2: ```suggestion
This element with the `menu` role only owns elements with the `menuitem`, `menuitemradio` and `menuitemcheckbox` role. These roles are all [required owned elements][] for `menu`. The element with the `none` role is not [owned by][] the `menu` because it is not [included in the accessibility tree][].
```

```html
<ul role="menu">
	<li role="none"></li>
	<li role="menuitem">Item 1</li>
	<li role="menuitemradio">Item 2</li>
	<div role="menuitemcheckbox">Item 3</div>
</ul>
```

#### Passed Example 5

This element with the `list` role only owns elements with an implicit `listitem` role. The `listitem` role is one of the [required owned elements][] for `list`.

```html
<div role="list">
	<li>Item 1</li>
</div>
```

#### Passed Example 6

This element with the `list` role only owns elements with the `listitem` role through the `aria-owns` attribute. The `listitem` role is one of the [required owned elements][] for `list`.

**Note:** This test case follows the definition of [owned by][]. If implemented differently, this definition could cause differences in outcome of this test case.

```html
<div role="list" aria-owns="id1"></div>
<div id="id1" role="listitem">Item 1</div>
```

#### Passed Example 7

> This element with the `list` role only owns elements with the `listitem` role, or elements with the `group` role, in which each element has the `listitem` role. The `listitem` role on its own, as well as in a `group` are in the [required owned elements][] for `list`.

Reviewer 2: ```suggestion
This element with the `list` role only owns elements with the `listitem` role, or elements with the `group` role, in which each element has the `listitem` role. Both the `listitem` role on its own, and the `group` role (when containing elements with the `listitem` role) are [required owned elements][] for `list`.
```

```html
<div role="list">
	<span role="listitem">Item 1</span>
	<div role="group">
		<span role="listitem">Item 2</span>
		<span role="listitem">Item 3</span>
	</div>
</div>
```

> ### Failed

Reviewer 2: I think we should add an example with an element owning one good and one bad role, to illustrate the "only" part of the expectation:
```html
<div role="list">
  <li>Item 1</li>
  <span role="link">Item 2</span>
</div>
```

#### Failed Example 1

> This element with the `list` role owns elements without the `listitem` role. The `listitem` role is one of the [required owned elements][] for `list`.

Reviewer 2: ```suggestion
This element with the `list` role does not owns an element without any of its [required owned elements][] (`listitem` or `group` containing `listitem`).
```
Because 
```html
<div role="list">
  <div role="group">
    <li>Foo</li>
  </div>
</div>
```
is matching the description (`list` owns elements without `listitem` role) but passes the rule.

```html
<div role="list" aria-owns="id1"></div>
<div id="id1" role="listitem">Item 1</div>
```

#### Passed example 10

Element with role `menu` has attribute `aria-busy` set to `true`.

```html
<ul role="menu" aria-busy="true">Loading</ul>
```

#### Passed example 11

Element `ul` with implicit role of `list` owns at least one instance of required owned element, in this case a `span` with `role` of `listitem`.

```html
<ul>
	<span role="listitem">Item 1</span>
	<li>Item 2</li> <!-- implicit role -->
</ul>
```

### Failed

#### Failed example 1

Element with role `list` is missing required owned element `listiem`.

```html
<ul role="list">
	<span>Item 1</span>
</ul>
```

#### Failed example 2

Element with role `tablist` is missing required owned element `tab`.

```html
<ol role="tablist">
	<li role="listitem">Item 1</li>
</ol>
```

#### Failed example 3

Element with role `grid` owns required owned element `rowgroup`, which is missing its required owned element `row`.

```html
<div role="grid">
	<div>
		<div role="rowgroup">
			<span>Item 1</span>
		</div>
		<div></div>
	</div>
</div>
```

#### Failed example 4

> The element with the semantic role of `list` owns the element with the role `tab` through the `aria-owns` property.

Reviewer: To make it clearer why the example fails, you can add something like 
", but does not own the required elements with semantic roles `group` or `listitem`" 
to the end of the sentence

---
id: bc4a75
name: ARIA required owned elements
rule_type: atomic
description: |
 This rule checks that each role has at least one of its required owned elements.
accessibility_requirements:
 wcag20:1.3.1: # Info and Relationships (A)
  forConformance: true
  failed: not satisfied
  passed: further testing needed
  inapplicable: further testing needed
input_aspects:
- DOM Tree
- CSS Styling
authors:
- Audrey Maniez
- Jey Nandakumar
---

## Applicability

> This rule applies to any HTML or SVG element that is [exposed to assistive technologies](#exposed-to-assistive-technologies) and has an explicit [semantic role](#semantic-role) that has [WAI-ARIA required owned elements](https://www.w3.org/TR/wai-aria/#mustContain), except when the owned element has an implicit semantic role that is identical to its explicit semantic role.

Reviewer 2: I do not really understand:
* why we only want to check element with explicit roles, not those with implicit roles.
* why there is the exception at the end of the applicability clause.

(that's more curiosity than real comment)

Author: Thanks for the above 2 comments. I was trying to line up everything similar to #255 and that caused some ambiguity. I am updating this to simple language.

## Expectation

The target element [owns](#owned-by) at least one instance of one of its [WAI-ARIA required owned element](https://www.w3.org/TR/wai-aria-1.1/#mustContain).

**Note:**

> - When a widget is missing required owned elements due to script execution or loading, authors [MUST]((w3.org/TR/wai-aria-1.1/#mustContain)) mark a containing element with `aria-busy` equal to `true`.

Reviewer 2: Shouldn't that note be an actual expectation? eg, "the target element either owns one of its `mustContain` or has `aria-busy=true`".

The way it’s written now, it seems that `<ul role="list" aria-busy=true></ul>` is failing the rule (no list item) while this is perfectly OK thanks to the `aria-busy` attribute.

Maybe we can also add an example with `aria-busy` to illustrate this.

> - When a widget is missing required owned elements due to script execution or loading, authors [MUST]((w3.org/TR/wai-aria-1.1/#mustContain)) mark a containing element with `aria-busy` equal to `true`.

Reviewer 2: The note does not apply only to widgets. The link you put says " For example, until a page is fully initialized and complete, an author could mark the document element as busy." I think the document is not a widget.

## Assumptions

> *There are currently no assumptions*

Reviewer 3: Normally I would ask to provide the assumption "Aria is being used to comply to WCAG" but we should now probably use a less generic text, (See https://github.com/act-rules/act-rules.github.io/issues/549).

	<li></li> <!-- implicit role -->
	<li role='menuitem'></li>
	<li role='menuitemradio'></li>
</ul>
```

#### Passed example 5

Required owned element does not have to be an immediate child of the role, but could be any descendant. (See accessibility support notes for details).

```html
<div role="table">
	<div>
		<div>
			<div role="row">
				<div role="cell">my cell</div>
				<div role="cell">and another cell</div>
			</div>
			<div role="row">
				<div role="cell">my cell</div>
				<div role="cell">and another cell</div>
			</div>
		</div>
	</div>
</div>
```

#### Passed example 6

Nested required owned elements.

```html
<div role="grid">
	<div role="rowgroup">
		<div role="row">
			<div role="gridcell">1</div>
			<div role="gridcell">Item 1</div>
			<div role="gridcell">2 Nos.</div>
		</div>
	</div>
	<div role="rowgroup">
		<div role="row">
			<div role="gridcell">1</div>
			<div role="gridcell">Item 1</div>
			<div role="gridcell">2 Nos.</div>
		</div>
	</div>
</div>
```

#### Passed example 7

Element with role `list` contains a required owned element which has an implicit role of `listitem`.

```html
<div role='list'>
   <li>
   </li>
</div>
```

#### Passed example 8

> Element with role `menubar` can have multiple required owned element (`menuitem` `menuitemcheckbox` and `menuitmeradio`) but only one is necessary.

Reviewer 2: I think it would be a bit better if the example contained 2 of the three possible elements but not all of them.

The reason is that an incorrect tool that would test for the presence of *all* of them would incorrectly fail on a test case with only some of them and thus we could detect such a problem.
It also makes it a bit easier (I think) for humans to understand "we don’t need all of them" if we show them that it's fine to have only some.

```html
<div role='menubar'>
   <div role='menuitem'>
   </div>
   <div role='menuitemcheckbox'>
   </div>
	 <div role='menuitmeradio'>
   </div>
</div>
```

#### Passed example 9

Element with role `list` contains a required `listitem`, the relationship between both elements is made possible through the `aria-owns` properties.

> **Note:** This test case follows the definition of [owned by](#owned-by) used in this rule. If implemented differently, this definition could cause differences in outcome for this test case.

Reviewer 2: This is the definition of "owned by" made by WAI ARIA, not just "used in this rule" (ie, it is not something we come up ourselves, it is the actual definition and if a browser/AT does not respect this definition, it's the browser/AT fault for not respecting WAI ARIA, not our fault for making weird definition).

I think this should rather link to the definition of "owned by" in WAI ARIA" (same as "Owned element" in the background Section).

Author: Because myself and @annethyme started authoring this rule (owned elements) and the [context rule](https://github.com/act-rules/act-rules.github.io/pull/255), after a long discussion of what definition to use for owned elements which happened here - https://github.com/act-rules/act-rules.github.io/issues/302, a decision was made to author our own. 

The definition is authored here - https://github.com/act-rules/act-rules.github.io/pull/255/files#diff-e8efc5e044aa522fbb04b4f072890fa3

```html
<div role="list" aria-owns="id1">
</div>

<div id="id1" role="listitem"></div>
```

#### Passed example 10

Element with role `grid` has at least one required owned element with role `rowgroup`.

```html
<div role="grid">
	<div>
		<div role="row">
			<span>Item 1</span>
		</div>
	<div>
	<div role="rowgroup">
		<div role="row">
			<span>Item 2</span>
		</div>
	</div>
</div>
```

### Failed

#### Failed example 1

Element with role `list` is missing required owned element `listitem`.

```html
<ul role='list'>
	<span>
	</span>
</ul>
```

#### Failed example 2

Element with role `tablist` is missing required owned element `tab`.

```html
<ol role='tablist'>
	<li role='listitem'>
	</li>
</ol>
```

#### Failed example 3

> Element with role `grid` is missing a required owned element `rowgroup`, even though the containing required owned element `row` is present.

Reviewer: The required owned elements for `grid` are `row` and `rowgroup`. Since the expectation requires just one instance of the required owned elements, this is a pass.

Author: Nice catch, I have updated this to something that will fail.

```html
<div role="grid">
	<div>
		<div role="row">
			<span>Item 1</span>
		</div>
	<div>
</div>
```

#### Failed example 4

The element with the semantic role of `list` owns the element with the role `tab` through the `aria-owns` property.

```html
<div role="list" aria-owns="id2">
</div>

<div id="id2" role="tab"></div>
```

> **Note:** This test case follows the definition of [owned by](#owned-by) used in this rule. If implemented differently, this definition could cause differences in outcome for this test case.

Reviewer 2: Same comments as for the note of Passed example 9.

> **Note:** This test case follows the definition of [owned by](#owned-by) used in this rule. If implemented differently, this definition could cause differences in outcome for this test case.

Reviewer 2: Additionally, the note on Passed example 9 is before the code, and here it's after the code. For the sake of consistency, I suggest we try and keep notes in the same place.

#### Failed Example 1

> This element with the `list` role does not owns an element without any of its [required owned elements][] (`listitem` or `group` containing `listitem`).

Reviewer 2: with
```suggestion
This element with the `list` role owns an element without any of its [required owned elements][] (`listitem` or `group` containing `listitem`).
```

</ul>
```

#### Inapplicable Example 6

> This element with the `combobox` role is accessible without owned elements.

Reviewer 2: Does "is accessible" means "is included in the accessibility tree"?
